src/os/bluestore/Btree2Allocator.cc

```diff
@@ -0,0 +1,615 @@
+    do {
+      // try spilled over or different direction if bucket index is out of bounds
+      if (bucket >= myTraits.num_buckets) {
+        if (dir < 0) {
+          // reached the bottom while going downhill,
+          // time to try spilled over extents
+          int64_t r = _spillover_allocate(size,
+            unit,
+            size,
+            0,
+            extents);
+          if (r > 0) {
+            allocated += r;
+            ceph_assert(size >= (uint64_t)r);
+            size -= r;
+            if (size == 0) {
+              return allocated;
+            }
+          } else {
+            ceph_assert(r == 0); // _spillover_allocate should
+                                 // handle errors internally
+          }
+        }
+        // change direction
+        dir = -dir;
+        bucket = dir < 0 ? bucket0 : bucket_center + 1; // See above on new bucket
+                                                        // selection rationales
+        ceph_assert(bucket < myTraits.num_buckets); // this should never happen
+        if (dir == dir0 ) {
+          // stop if both directions already attempted
+          return -ENOSPC;
+        }
+      }
+      rs_tree = &range_size_set[bucket];
+      rs_p = _pick_block(dir, rs_tree, size);
+      bucket += dir; // this might wrap over zero and trigger direction
+                     //  change on the next loop iteration.
+    } while (rs_p == rs_tree->end());
+  }
+
+found:
+  if (rs_p != rs_tree->end()) {
+    auto o = rs_p->start;
+    auto l = std::min(size, rs_p->length());
+    auto rt_p = range_tree.find(o);
+    ceph_assert(rt_p != range_tree.end());
+    _remove_from_tree(rs_tree, rs_p, rt_p, o, o + l);
+    extents->emplace_back(o, l);
+    allocated += l;
+    return allocated;
+  }
+  // should never get here
+  ceph_assert(false);
+  return -EFAULT;
+}
+
+Btree2Allocator::range_size_tree_t::iterator
+Btree2Allocator::_pick_block(int dir,
+                              Btree2Allocator::range_size_tree_t* tree,
+                              uint64_t size)
+{
+  if (dir < 0) {
+    // use the largest available chunk from 'shorter chunks' buckets
+    // given the preliminary check before the function's call
+    // we should get the result in a single step
+    auto p = tree->end();
+    if (p != tree->begin()) {
+      --p;
+      return p;
+    }
+  } else if (dir > 0) {
+    auto p = tree->begin();
+    if (p != tree->end()) {
+      return p;
+    }
+  } else if (tree->size()) {
+    //For the 'best matching' bucket
+    //look for a chunk with the best size matching.
+    // Start checking from the last item to
+    // avoid lookup if possible
+    auto p = tree->end();
+    --p;
+    if (p->length() > size) {
+      p = tree->lower_bound(range_seg_t{ 0, size });
+      ceph_assert(p != tree->end());
+    }
+    return p;
+  }
+  return tree->end();
+}
+
+void Btree2Allocator::_remove_from_tree(uint64_t start, uint64_t size)
+{
```

Comment:
Once we create the Bitmap side of hybrid, there is no path when we decide that it is no longer needed?

Reply:
No. That's an original hybrid allocator behavior. And I doubt we really need that in real production before we have some working defragmentation.